pkg/client/introduce/client.go

```diff
@@ -0,0 +1,147 @@
+func (c *Client) SendProposal(dest1, dest2 *service.Destination) error {
+	return c.sendProposal(options{
+		Destinations: []*service.Destination{dest1, dest2},
+	})
+}
+
+// SendProposalWithInvitation sends proposal to introducee
+// executes when the client has a public Invitation
+func (c *Client) SendProposalWithInvitation(inv *didexchange.Invitation, dest *service.Destination) error {
+	return c.sendProposal(options{
+		Invitation:   inv,
+		Destinations: []*service.Destination{dest},
+	})
+}
+
+// SendRequest sends request
+// sending a request means that introducee is willing to share its invitation
+func (c *Client) SendRequest(dest *service.Destination) error {
+	return c.handleOutbound(&introduce.Request{
+		Type: introduce.RequestMsgType,
+		ID:   uuid.New().String(),
+	})
+}
+
+// HandleRequest executes after receiving a request (the client does not have a public Invitation)
+func (c *Client) HandleRequest(dest1, dest2 *service.Destination) error {
+	return c.sendProposal(options{
+		Destinations: []*service.Destination{dest1, dest2},
+	})
+}
+
+// HandleRequestWithInvitation executes after receiving a request when the client has a public Invitation
+func (c *Client) HandleRequestWithInvitation(inv *didexchange.Invitation, dest *service.Destination) error {
+	return c.sendProposal(options{
+		Invitation:   inv,
+		Destinations: []*service.Destination{dest},
+	})
+}
+
+func (c *Client) handleOutbound(msg interface{}) error {
+	payload, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	didMsg, err := service.NewDIDCommMsg(payload)
+	if err != nil {
+		return err
+	}
+
+	return c.service.HandleOutbound(didMsg, &service.Destination{})
+}
+
+// options keeps the information needed for sending a proposal
+type options struct {
+	// Invitation must be set when we have a public Invitation
+	Invitation *didexchange.Invitation
+	// Destinations contain one or two elements
+	// one element - for a public Invitation, otherwise two elements
+	Destinations []*service.Destination
+}
+
+// sendProposal sends proposal
+// the function may receive an `Invitation` or <nil> (e.g options.Invitation)
+// if the invitation is not <nil> it means we have a public invitation.
+// Destinations (e.g options.Destinations) should have one or two values, if destinations have only one value it means
+// that it is public invitation (an invitation `options.Invitation` is required), otherwise if destinations have two
+// values, it means we do not have a public invitation (we will wait for it from one of introducees)
+// Usage:
+//  Sends public invitation to the destination
+// 	sendProposal(&options{
+// 		Invitation: invitation,
+// 		Destinations: []*service.Destination{
+// 			destination,
+// 		},
+// 	})
+// 	Forwards an invitation from one introducee to another
+// 	sendProposal(&options{
+// 		Destinations: []*service.Destination{
+// 			destination1,
+// 			destination2,
+// 		},
+// 	})
+func (c *Client) sendProposal(_ options) error {
```

Comment:
why the ununsed argument? what happened to thread id?

Reply:
I added thread to Proposal structure bc I did not know how to handle it without thread, now I came up with some idea (it will be done in the next PR). The idea is that the client will keep an agent invitation, any time we want to accept the proposal this invitation will be passing to the Continue function, it allows us to proceed working with protocol without handling additional logic(I mean handle specific msg type)